### ISSUE TITLE
API Builder Not Working

### ISSUE THREAD

**Issue by NNewar** [Axway-API-Management-Plus/apigateway-openlogging-elk#36]
**Axway APIM-ELK version**
I have setup the environment using the docker-compose image provided in the RC-1.0.0 version

**API-Manager Version**
APIM - 7.6.2

**Question**
I am unable to fetch the information through the API Builder, however, I can see my logs inside the Kibana Dashboard. I have also used the Elastic Search APIs just to confirm if the transaction ID information is available or not.

Also, I am unable to access the API Builder Console. I am getting 404 Not Found Error.

![Kibana-Dashboard](https://user-images.githubusercontent.com/71384302/93332951-40624980-f840-11ea-8409-92ccde608b06.PNG)
![Postman-APIBuilder-Call](https://user-images.githubusercontent.com/71384302/93332959-41937680-f840-11ea-9f74-cb8c61517ac0.PNG)
![Postman-Elasticsearch-Call](https://user-images.githubusercontent.com/71384302/93332962-422c0d00-f840-11ea-9ba8-a58bc3a17e4d.PNG)

**What I've tried so far**
API Builder Console Error -
I have tried allowed a list of IP addresses and disabled the allowed host configuration inside the conf/default.js, still unable to access the API Builder console.

API Builder APIs -
I have tried to change the host name of the Elasticsearch inside the "conf/elasticsearch.js" file.

**Additional information**
I have tried sending the request from Postman and even setup the API Node Manager configuration to call the API Builder APIs. I am still unable to see the information.

**Comment by cwiechmann** [comment]
Hey @NNewar, Is the API-Builder Docker-Container running? What is shown, when you make a docker logs on this container? 

FYI: 
_ - The API-Builder console is not part of the generated Docker-Image._
_ - The events are processed anyway by Logstash, but some additional Lookup-Information are then not attached_

I made a fix, which will make sure, that this message will appear in the Traffic-Monitor if the Lookup by Logstash fails:
![image](https://user-images.githubusercontent.com/16292002/93333735-0edb8500-f824-11ea-83a6-7cee78d01f4a.png)

**Comment by NNewar** [comment]
Hey @cwiechmann, 

The API Builder Docker-Container is running:

![Docker-Containers](https://user-images.githubusercontent.com/71384302/93338012-8078fa80-f847-11ea-98c5-197124f6403e.PNG)

And, I can access the API Builder through browser as well:

![APIBuilderBrowser](https://user-images.githubusercontent.com/71384302/93338132-a56d6d80-f847-11ea-9371-a75a687907b8.PNG)

I have checked the logs for the docker container and I can see that the requests are coming to the API Builder:

[APIBuilder-Docker.log](https://github.com/Axway-API-Management-Plus/apigateway-openlogging-elk/files/5232192/APIBuilder-Docker.log)

**Comment by cwiechmann** [comment]
So, you mean the API-Gateway Traffic-Monitor is not showing up the requests?

**Comment by NNewar** [comment]
Yes, Traffic-Monitor is not showing up the requests

**Comment by cwiechmann** [comment]
Okay. The most important index for that information to come up is the apigw-traffic-summary-* index. 

Question: Have you deleted the index-templates and indices that might have exists before in Elasticsearch?

I recommend to enable debug for API-Builder which you can do in the docker-compose file (there is a comment). 
After you have restarted, best is to filter for a known Transaction-ID to reduce the overhead and refresh the Traffic-Screen. Then you see what API-Builder is doing. Have a look for these lines:
```
Query Elasticsearch (elasticsearch.1) route: ["javascript.1"]
search (elasticsearch.1) Using elastic search query body: {"index":"apigw-traffic-summary-*","body":{"query":{"bool":{"must":[{"match":{"correlationId":"e2ff615f85043eab35b77de1"}},{"term":{"processInfo.serviceId":"instance-1"}},{"exists":{"field":"http"}}]}}},"size":"1000","sort":"@timestamp:desc"}
Query Elasticsearch (elasticsearch.1) route: ["javascript.1"]
Query Elasticsearch (elasticsearch.1) writing {"body":{"took":13,"timed_out":false,"_shards":{"total":5,"successful":5,"skipped":0,"failed":0},"hits":{"total":{"value":1,"relation":"eq"},"max_score":null,"hits":[{"_index":"apigw-traffic-summary-000001","_type":"_doc","_id":"e2ff615f85043eab35b77de1","_score":null,"_source":{"input":{"type":"log"},"finalStatus":"Fail","processInfo":{"serviceId":"instance-1","groupId":"group-2","gatewayName":"API-Gateway","hostname":"api-env","gatewayRegion":"N/A","groupName":"QuickStart Group","version":"7.7.20200730"},"@timestamp":"2020-09-16T12:07:28.204Z","logtype":"openlog","@version":"1","type":"summaryIndex","http":{"remoteName":"192.168.65.1","uri":"/api/v1/airports/fds","statusText":"Internal Server Error","localAddr":"192.168.65.133","status":500,"bytesReceived":368,"bytesSent":587,"localPort":"8065","wafStatus":0,"method":"GET","remotePort":"53823","remoteAddr":"192.168.65.1","authSubjectId":"Pass Through"},"correlationId":"e2ff615f85043eab35b77de1","duration":30061,"protocol":"https","protocolSrc":"8065","serviceContext":{"method":"get-code","monitor":true,"routingPolicy":"Custom routing policy","apiOrg":"API Development","responsePolicy":"Custom response policy","apiVersion":"1.0.1","service":"Airports API","apiState":"unpublished","apiDeprecated":"false","apiSecurity":"Pass Through","requestPolicy":"Custom request policy","backendBasePath":"http://192.168.65.1:8080","faulthandlerPolicy":"N/A","status":"failure"},"status":"failure"},"sort":[1600258048204]}]}},"statusCode":200,"headers":{"content-type":"application/json; charset=UTF-8","content-length":"1474","access-control-allow-credentials":"true"},"warnings":null,"meta":{"context":null,"request":{"params":{"method":"POST","path":"/apigw-traffic-summary-*/_search","body":"{\"query\":{\"bool\":{\"must\":[{\"match\":{\"correlationId\":\"e2ff615f85043eab35b77de1\"}},{\"term\":{\"processInfo.serviceId\":\"instance-1\"}},{\"exists\":{\"field\":\"http\"}}]}}}","querystring":"size=1000&sort=%40timestamp%3Adesc","headers":{"user-agent":"elasticsearch-js/7.9.0 (linux 3.10.0-1127.18.2.el7.x86_64-x64; Node.js v12.18.3)","content-type":"application/json","content-length":"160"},"timeout":60000},"options":{"ignore":[404],"maxRetries":3,"warnings":null},"id":6},"name":"apibuilder4elastic","connection":{"url":"https://elasticsearch1:9200/","id":"https://elasticsearch1:9200/","headers":{},"deadCount":0,"resurrectTimeout":0,"_openRequests":0,"status":"alive","roles":{"master":true,"data":true,"ingest":true,"ml":false}},"attempts":0,"aborted":false}} to ctx as: $.elasticsearch.result
```

You can take the query body from the log line and run the query manually:
```
GET /apigw-traffic-summary-*/_search
{
  "query": {
    "bool": {
      "must": [
        {
          "match": {
            "correlationId": "e2ff615f85043eab35b77de1"
          }
        },
        {
          "term": {
            "processInfo.serviceId": "instance-1"
          }
        },
        {
          "exists": {
            "field": "http"
          }
        }
      ]
    }
  }
}
```
If this query doesn't return anything, then you can remove one by one of the constraints.

**Comment by cwiechmann** [comment]
@NNewar, do you made any progress and still need help or can I close this issue?

**Comment by taranwar** [comment]
Hi @cwiechmann , 

Nischal is on leave today so he should be able to test and confirm tomorrow. 

Thanks 
Tarique

**Comment by NNewar** [comment]
Hello @cwiechmann,

I have found out the issue, why the API calls from Postman was not working. I enabled the API Builder debug logging by uncommenting the line inside the docker-compose file.

- After enabling it, I sent a request from Postman and found out that the API is expecting the Cookie header (which is provided in the API Gateway Manager requests).

![image](https://user-images.githubusercontent.com/71384302/93570628-ea60e380-f9b0-11ea-8314-0402fc116940.png)

- I copied the Cookie header value from my local API Gateway Manager and then added it to my Postman Call.

![image](https://user-images.githubusercontent.com/71384302/93570713-0bc1cf80-f9b1-11ea-92da-e4ce8bfc3aa3.png)

- Then I sent the request from my Postman and got the response:

![image](https://user-images.githubusercontent.com/71384302/93570806-33189c80-f9b1-11ea-9594-65aa4d24be9f.png)

I am now sure that the docker container is properly configured and API Builder is providing the information.

I will now add the custom Elasticsearch policy inside the ANM and will test the complete solution.

Thank you for your help.
